### PR TITLE
503 get unique product id latest UID

### LINKED ISSUE #503: DigitalShadow: GetUniqueProductID gets wrong ID
*Related to Digital Shadow: #297*

**Context:**
One item (with one AID) can be recorded multiple times at one asset, meaning that the combination of assetID and AID isn't unique. This is correctly represented by the timescaleDB ddl. If a product is passing multiple times at one asset (maybe because of rework) the GetUniqueProductID function should always return the latest UID.

**Problem:**
The current func `GetUniqueProductID` always returns the oldest and not the youngest UID. This is because the SQL query doesn't accout for the possibility of non-unique assetID and AID combinations and returns multiple UID (default in ascending order).

**Solution:**
Order the UID's by timestamp and select only the latest one.

**Disclaimer**
An AID is likely to be recorded multiple times (e.g. in case of scanning twice). Users have to make shure to only send the intended uniqueProduct MQTT messages.

### PR DESCRIPTION
**Pull request by larswitte**
Fixes #503

### CODE REVIEW

File: golang/internal/cache.go [hunk 1]
```diff
@@ -561,62 +561,6 @@ func StoreAssetIDToCache(customerID string, location string, assetID string, DBa
 
 	b := strconv.Itoa(DBassetID)
 
-	err := rdb.Set(ctx, key, b, dataExpiration).Err()
-	if err != nil {
-		zap.S().Errorf("redis failed")
-		return
-	}
-}
-
```
Comment (larswitte) [inline golang/internal/cache.go:570]: I didn't delete this part, I just deleted the caching parts. Is it just a displaying wrongly or some underlying issue?
Please check @JeremyTheocharis.

File: golang/internal/cache.go [hunk 2]
```diff
@@ -561,62 +561,6 @@ func StoreAssetIDToCache(customerID string, location string, assetID string, DBa
...
-// StoreUniqueProductIDToCache stores uniqueProductID to cache
-func StoreUniqueProductIDToCache(aid string, DBassetID int, uid int) {
-	if rdb == nil { // only the case during tests
-		//zap.S().Errorf("rdb == nil")
-		return
-	}
-
-	key := fmt.Sprintf("getUniqueProductID-%s-%d", aid, DBassetID)
```
Comment (larswitte) [inline golang/internal/cache.go:611]: I also didn't edit this part...

File: golang/internal/cache.go [hunk 3]
```diff
@@ -561,62 +561,6 @@ func StoreAssetIDToCache(customerID string, location string, assetID string, DBa
...
-
-	if uid == 0 {
-		// zap.S().Debugf("input is empty. aborting storing into database.", key)
-		return
-	}
-
-	b := strconv.Itoa(uid)
-
 	err := rdb.Set(ctx, key, b, dataExpiration).Err()
 	if err != nil {
```
Comment (larswitte) [inline golang/internal/cache.go:565]: I deleted this part.

File: golang/internal/cache.go (hunk 1 above)
Comment (larswitte) [reply golang/internal/cache.go:570]: The red marking seems to be 6 lines to high.

**Comment by larswitte** [comment]
I checked it with my node red test flow: worked!

**Review by JeremyTheocharis — COMMENTED** [review]
@Scarjit 

See also other PR

**Comment by Scarjit** [comment]
Integrated in #500 (https://github.com/Scarjit/united-manufacturing-hub/commit/4c2f9854c9ea8bf4eb1b1dc1fdf8f260547e4ae8)

**Pull request closed without merging**